python/rapidsmpf/rapidsmpf/integrations/dask/shuffler.py

```diff
@@ -22,24 +23,57 @@
 if TYPE_CHECKING:
     from collections.abc import Callable, MutableMapping, Sequence
 
-    from distributed import Worker
+    from distributed import Client, Worker
 
 
-_shuffle_counter: int = 0
+# Set of available shuffle IDs
+_shuffle_id_vacancy: set[int] = set(range(Shuffler.max_concurrent_shuffles))
+_shuffle_id_vacancy_lock: threading.Lock = threading.Lock()
 
 
-def get_shuffle_id() -> int:
+def _get_new_shuffle_id(client: Client) -> int:
     """
-    Return the unique id for a new shuffle.
+    Get a new available shuffle ID.
+
+    Since RapidsMPF only supports a limited number of shuffler instances at
+    any given time, this function maintains a shared pool of shuffle IDs.
+
+    If no IDs are available locally, it queries all workers for IDs in use,
+    updates the vacancy set accordingly, and retries. If all IDs are in use
+    across the cluster, an error is raised.
+
+    Parameters
+    ----------
+    client
+        A Dask distributed client used to query workers for active shuffle IDs.
 
     Returns
     -------
-    The enumerated integer id for the current shuffle.
+    A unique shuffle ID not currently in use.
+
+    Raises
+    ------
+    ValueError
+        If all shuffle IDs are currently in use across the cluster.
     """
-    global _shuffle_counter  # noqa: PLW0603
+    with _shuffle_id_vacancy_lock:
+        if not _shuffle_id_vacancy:
+
+            def get_vacancy_ids(dask_worker: Worker) -> set[int]:
+                ctx = get_worker_context(dask_worker)
+                with ctx.lock:
+                    return set(ctx.shufflers.keys())
+
+            _shuffle_id_vacancy.update(range(Shuffler.max_concurrent_shuffles))
+            _shuffle_id_vacancy.difference_update(*client.run(get_vacancy_ids).values())
 
-    _shuffle_counter += 1
-    return _shuffle_counter
+            if not _shuffle_id_vacancy:
+                raise ValueError(
+                    f"Cannot shuffle more than {Shuffler.max_concurrent_shuffles} "
+                    "times in a single Dask compute."
+                )
+
+        return _shuffle_id_vacancy.pop()
```

Comment:
Do we ever (want to) deregister shuffler IDs after they're done?

Reply:
Yes, to support more than 256 shuffles in total, we need to release IDs.

Reply:
Sorry, I think I wasn't clear, I mean that I _think_ they are not released _after use_, here the ID is "given" to the caller, but when the caller doesn't use it anymore should it be put back into `_shuffle_id_vacancy` somehow? I didn't see anything of that sort, how is that handled currently besides having to ask workers for [occupied IDs](https://github.com/rapidsai/rapidsmpf/pull/254/files#diff-f6d20cdd9996b100c878b44518f46cfcda72fec240c6420adfe590c66ea96cfaR64-R67)? Is there an additional mechanism for that, and if not, should we have that?

Reply:
No, we don't have a mechanism for that. This is way the client has to ask the workers. Notice, all workers must have completed the shuffle before we can reuse that ID. 

Since 256 is a high number, I think it is fine to call the workers once all of the 256 ID has been used.

Reply:
Makes sense, thanks for clarifying.